### PR TITLE
Add dependency installation step to support deploying config as code from worker

### PR DESCRIPTION
**Pull request by ameyakhare**
Resolves TSK-116

## Description
Adds a step to install dependencies (once per project root) during a deploy so that the worker can properly import user code to extract task and view configs.

## Test plan
Deploying locally works when I remove the node_modules directory and it doesn't reinstall if the directory exists.
Deploying locally is significantly faster now, despite having numerous files.

### CODE REVIEW

File: pkg/deploy/discover/parser/node/parser.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
+var path = require("node:path");
```
Comment (leeweisberger) [inline pkg/deploy/discover/parser/node/parser.js:1]: Can you output a comment that this is autogenned please

File: pkg/deploy/discover/code_task_discover.go [hunk 2]
```diff
@@ -142,6 +143,7 @@ func (c *CodeTaskDiscoverer) parseNodeDefinitions(ctx context.Context, file stri
 
 	parsedConfigs, err := extractConfigs(compiledJSPath)
 	if err != nil {
+		fmt.Println(err)
```
Comment (leeweisberger) [inline pkg/deploy/discover/code_task_discover.go:146]: You mean to print this?

File: pkg/deploy/discover/build.go [hunk 3]
```diff
@@ -11,14 +12,52 @@ import (
...
 
+//go:generate npx -p typescript tsc parser/node/parser.ts
+//go:embed parser/node/parser.js
+var nodeParserScript []byte
+
 const outputDir = ".airplane"
 
+func maybeInstallDependencies(rootDir string) error {
+	if !fsx.Exists(path.Join(rootDir, "node_modules")) {
+		// node_modules doesn't exist, we may need to install dependencies
```
Comment (leeweisberger) [inline pkg/deploy/discover/build.go:30]: What if their node_modules exists but is not up to date? Like I added a new dep in my code and package.json but didn't run npm install?

File: pkg/deploy/discover/build.go [hunk 4]
```diff
@@ -11,14 +12,52 @@ import (
...
+				if isYarnBerry, _ := build.IsYarnBerry(packageJSONPath); isYarnBerry {
+					cmd = exec.Command("yarn", "install")
+				} else {
+					cmd = exec.Command("yarn", "install", "--non-interactive", "--production", "--frozen-lockfile")
+				}
+
+			} else if fsx.Exists(path.Join(rootDir, "package-lock.json")) {
+				cmd = exec.Command("npm", "ci", "--production")
+			} else {
+				cmd = exec.Command("npm", "install", "--production")
```
Comment (leeweisberger) [inline pkg/deploy/discover/build.go:46]: Why are these different?

File: pkg/deploy/discover/code_task_discover.go [hunk 5]
```diff
@@ -120,6 +117,10 @@ func (c *CodeTaskDiscoverer) parseNodeDefinitions(ctx context.Context, file stri
 		return nil, errors.Wrap(err, "unable to interpret task path metadata")
 	}
 
+	if err := maybeInstallDependencies(pm.RootDir); err != nil {
+		return nil, err
+	}
```
Comment (leeweisberger) [inline pkg/deploy/discover/code_task_discover.go:122]: So the theory is we should not be running this if the deploy is local. We will run it in the server. 

We are going to install deps twice in that case right? Once during discovery and then a second time during build. That's going to slow things down a bit.

Also we aren't caching deps here right?

**Review by leeweisberger — COMMENTED** [review]
Want to make sure I understand correctly / we understand the perf implications here

File: pkg/deploy/discover/build.go (hunk 3 above)
Comment (ameyakhare) [reply pkg/deploy/discover/build.go:30]: if they are developing locally, we won't override their node_modules. unsure if that makes more sense or if we should always install when attempting to discover their tasks. I imagine the later option might be a little slow during local development and isn't something we do right now anyways

File: pkg/deploy/discover/build.go (hunk 4 above)
Comment (ameyakhare) [reply pkg/deploy/discover/build.go:46]: part of this has been copied from the build process (but not enough of it was similar enough to refactor and reuse)

File: pkg/deploy/discover/code_task_discover.go (hunk 5 above)
Comment (ameyakhare) [reply pkg/deploy/discover/code_task_discover.go:122]: I believe npm/yarn caches dependencies in a global system cache. downloading a particular dependency would be slow until the cache warms up

File: pkg/deploy/discover/parser/node/parser.js (hunk 1 above)
Comment (ameyakhare) [reply pkg/deploy/discover/parser/node/parser.js]: I've removed this for now and will do it in another diff

**Review by leeweisberger — APPROVED** [review]
Approving, but please refactor the install logic out and reuse in node.go

Also add some more logging.

I'm still worried about adding so much latency installing dependencies twice without caching. We will probably have to come up with something better

File: pkg/deploy/discover/code_task_discover.go (hunk 5 above)
Comment (leeweisberger) [reply pkg/deploy/discover/code_task_discover.go:122]: These worker pods are not very long lived, so we can't really count on that

**Comment by ameyakhare** [comment]
I'm abandoning this in favor of the new deployments work instead

**Pull request closed without merging**